content/blog/2022/2022-06-30-configuring-cors-with-spring.md

```diff
@@ -0,0 +1,430 @@
+    <artifactId>spring-boot-starter-security</artifactId>
+  </dependency>
+````
+Gradle:
+````groovy
+  implementation 'org.springframework.boot:spring-boot-starter-security'
+````
+
+#### Spring Security applied to a Spring Boot application
+
+Spring security by default protects every endpoint. However this would not hold true in case of preflight requests since **the purpose of an OPTIONS request 
+is only to establish if the original request would go through**. To make Spring Security bypass preflight requests we need to add **http.cors()** as shown:
+
+````java
+@Configuration
+@EnableConfigurationProperties(BasicAuthConfigProperties.class)
+@EnableWebSecurity
+public class SecurityConfiguration extends WebSecurityConfigurerAdapter {
+
+    private final BasicAuthConfigProperties basicAuth;
+
+    public SecurityConfiguration(BasicAuthConfigProperties basicAuth) {
+        this.basicAuth = basicAuth;
+    }
+
+    protected void configure(HttpSecurity http) throws Exception {
+        http.cors();
+    }
+}
+
+````
+
+To setup additional CORS configuration with Spring Security after bypassing pre-flight requests, we can
+- **Configure CORS using @CrossOrigin**
+
+````java
+@CrossOrigin(maxAge = 3600, allowCredentials = "true")
+@RestController
+@RequestMapping("cors-library/managed/books")
+public class LibraryController {
+
+    private static final Logger log = LoggerFactory.getLogger(LibraryController.class);
+
+    private final LibraryService libraryService;
+
+    public LibraryController(LibraryService libraryService) {
+        this.libraryService = libraryService;
+    }
+
+    @CrossOrigin(origins = "http://localhost:4200", allowedHeaders = {"Requestor-Type", "Authorization"}, exposedHeaders = "X-Get-Header")
+    @GetMapping
+    public ResponseEntity<List<BookDto>> getBooks(@RequestParam String type) {
+        HttpHeaders headers = new HttpHeaders();
+        headers.set("X-Get-Header", "ExampleHeader");
+        return ResponseEntity.ok().headers(headers).body(libraryService.getAllBooks(type));
+    }
+}
+````
+
+- OR **Create a CorsConfigurationSource bean**
+
+````java
+    @Bean
+    CorsConfigurationSource corsConfigurationSource() {
+            CorsConfiguration configuration = new CorsConfiguration();
+            configuration.setAllowedOrigins(Arrays.asList("http://localhost:4200"));
+            configuration.setAllowedMethods(Arrays.asList("GET","POST","PATCH", "PUT", "DELETE", "OPTIONS", "HEAD"));
+            configuration.setAllowCredentials(true);
+            configuration.setAllowedHeaders(Arrays.asList("Authorization", "Requestor-Type"));
+            configuration.setExposedHeaders(Arrays.asList("X-Get-Header"));
+            configuration.setMaxAge(3600L);
+            UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
+            source.registerCorsConfiguration("/**", configuration);
+            return source;
+    }
+````
+
+##### Spring Security applied to a Spring Webflux
```

Comment:
The header-depth is a bit random here. I have changed it in a couple places. Please have a look if that makes sense.

Reply:
Thank you for that. The headers look better structured now. :)